### PR TITLE
docs/features/wired-mesh: update to role-based configuration

### LINKED ISSUE #2474: docs: change Commandline reference of Wired mesh needs update
After #2372 got merged the Commandline Sections in the docs seems outdated (same goes for the wiki). As far as I understand it configuration should be done by editing the gluon.*. section and running gluon-reconfigure.

https://github.com/freifunk-gluon/gluon/blob/master/docs/features/wired-mesh.rst#configuration

### PR DESCRIPTION
**Pull request by AiyionPrime**
> - explain what happens on gluon-reconfigure
> - show workflow to alter the wired network config
> - update examples
> - update 'has changed in' section
> 
> resolves #2474 

@herbetom lets take this as a first draft. Let me know what you're missing or'd like to have changed.

### CODE REVIEW

File: docs/features/wired-mesh.rst [hunk 1]
```diff
@@ -56,32 +56,54 @@ the ``mesh`` role to the ``interfaces.*.default_roles`` options in site.conf.
...
+It is furthermore possible to make use of VLANs.
+The following section would create a VLAN with id 8 on ``eth0`` and emit the mesh network on it::
+
+  config interface 'iface_lan_vlan8'
+	option name 'eth0.8'
+	list role 'mesh'
+
+Other VLAN-interfaces could be configured on the same parent interface in order to have all three roles avaliable on ``eth0`` without having them interfere with each other.
+This feature comes in especially handy in the persistent configuration of virtual machines as offloader for bigger installations.
+
```
Comment (herbetom) [inline docs/features/wired-mesh.rst:104]:
```suggestion

It is possible to configure multiple interfaces with the same role, but currently all those interfaces will be bridged without any isolation. 
Mesh nodes that are connected to one of the mesh interfaces of a node can see each other directly.

```

sadly it doesn't look like #2484 (or something like it) will be merged in time for 2022.1

but idk, maybe it's better to leave this "disclaimer" away

Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:104]: I agree on the other two points, but looking at the fact that we intend to change the behaviour, I'd avoid explicitly documenting that there's no isolation.

File: docs/features/wired-mesh.rst [hunk 2]
```diff
@@ -56,32 +56,54 @@ the ``mesh`` role to the ``interfaces.*.default_roles`` options in site.conf.
 Commandline
 ===========
 
+Since version 2022.1, the wired network configuration is rebuilt from ``/etc/config/gluon`` on each ``gluon-reconfigure``, and therefore overwritten at least on every update of the node.
+There are currently three roles available for interfaces: ``client``, ``mesh`` and ``uplink``.
+
+Be careful: the ``client`` role must not be mixed with the other two.
```
[the remark this answers is not in the document]
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst]: You are right; I hope it's clearer now.

File: docs/features/wired-mesh.rst [hunk 3]
```diff
@@ -56,32 +56,54 @@ the ``mesh`` role to the ``interfaces.*.default_roles`` options in site.conf.
...
 Disable Mesh-on-LAN::
 
-  uci set network.mesh_lan.disabled=1
-  for ifname in $(cat /lib/gluon/core/sysconfig/lan_ifname); do
-    uci add_list network.client.ifname=$ifname
-  done
-  uci commit network
+  config interface 'iface_lan'
+	option name '/lan'
+	list role 'client'
```
[the remark this answers is not in the document]
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst]: I thought showing the config file would be easier to understand, but looking at the (slightly repetitive) mesh section now, just looks like a show of, of how clean the configuration can become ;)
Thanks for the hint.
I think this is resolved in 8e736b8c7d920a995ba4cc582b740f9cb10fdfcd.

File: docs/features/wired-mesh.rst [hunk 4]
```diff
@@ -56,32 +56,63 @@ the ``mesh`` role to the ``interfaces.*.default_roles`` options in site.conf.
...
+
+Be careful: When the client role is assigned to an interface at the same time as other roles (like 'client', 'mesh' in the Mesh-on-LAN example below), the other roles take precedence (enabling mesh, but not client in said example).
+
+The setup/config-mode interface is every interface with the role ``client``, which makes removing it from interfaces not only unnecessary, but generally unrecommended.
+
+In order to make persistent changes to the routers configuration it's necessary to:
+
+* change the sections in ``/etc/config/gluon`` e.g. using uci (see examples below)
+* call ``gluon-reconfigure`` to regenerate ``/etc/config/network`` afterwords
+* restart networking using either ``/etc/init.d/network restart`` or cleaner ``reboot`` the device
```
[the remark this answers is not in the document]
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst]: applied in 9ff834dd1f5a36d9f916bd1aa0527b5372ed2f82

File: docs/features/wired-mesh.rst [hunk 5]
```diff
@@ -56,32 +56,63 @@ the ``mesh`` role to the ``interfaces.*.default_roles`` options in site.conf.
...
+Furthermore it is possible to make use of VLANs.
+The following statements would create a VLAN with id 8 on ``eth0`` and emit the mesh network on it::
+
+  uci set gluon.iface_lan_vlan8=interface
+  uci set gluon.iface_lan_vlan8.name='eth0.8'
+  uci add_list gluon.iface_lan_vlan8.role='mesh'
+  uci commit gluon
+
+Other VLAN-interfaces could be configured on the same parent interface in order to have all three roles avaliable on ``eth0`` without having them interfere with each other.
+This feature comes in especially handy in the persistent configuration of virtual machines as offloader for bigger installations.
```
[the remark this answers is not in the document]
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst]: applied in 9ff834dd1f5a36d9f916bd1aa0527b5372ed2f82

File: docs/features/wired-mesh.rst [hunk 6]
```diff
@@ -56,32 +56,63 @@ the ``mesh`` role to the ``interfaces.*.default_roles`` options in site.conf.
...
+There are currently three roles available: ``client``, ``mesh`` and ``uplink``.
+
+Be careful: When the client role is assigned to an interface at the same time as other roles (like 'client', 'mesh' in the Mesh-on-LAN example below), the other roles take precedence (enabling mesh, but not client in said example).
+
+The setup/config-mode interface is every interface with the role ``client``, which makes removing it from interfaces not only unnecessary, but generally unrecommended.
+
+In order to make persistent changes to the routers configuration it's necessary to:
+
+* change the sections in ``/etc/config/gluon`` e.g. using uci (see examples below)
+* call ``gluon-reconfigure`` to regenerate ``/etc/config/network`` afterwords
```
[the remark this answers is not in the document]
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst]: applied in 9ff834dd1f5a36d9f916bd1aa0527b5372ed2f82

File: docs/features/wired-mesh.rst [hunk 7]
```diff
@@ -56,32 +56,63 @@ the ``mesh`` role to the ``interfaces.*.default_roles`` options in site.conf.
...
+
+  uci set gluon.iface_lan_vlan8=interface
+  uci set gluon.iface_lan_vlan8.name='eth0.8'
+  uci add_list gluon.iface_lan_vlan8.role='mesh'
+  uci commit gluon
+
+Other VLAN-interfaces could be configured on the same parent interface in order to have all three roles avaliable on ``eth0`` without having them interfere with each other.
+This feature comes in especially handy for the persistent configuration of virtual machines as offloader for bigger installations.
+
+It is not enough to ``reboot`` the device, calling ``gluon-reconfigure`` before is mandatory in order for the changes to take effect.
```
Comment (herbetom) [inline docs/features/wired-mesh.rst:114]: It's a bit redundant, but maybe it helps people, because it is quite important to know, idk

File: docs/features/wired-mesh.rst [hunk 8]
```diff
@@ -56,32 +56,63 @@ the ``mesh`` role to the ``interfaces.*.default_roles`` options in site.conf.
...
 Disable Mesh-on-LAN::
 
-  uci set network.mesh_lan.disabled=1
-  for ifname in $(cat /lib/gluon/core/sysconfig/lan_ifname); do
-    uci add_list network.client.ifname=$ifname
-  done
-  uci commit network
+  uci del_list gluon.iface_lan.role='mesh'
+  uci commit gluon
+
```
Comment (herbetom) [inline docs/features/wired-mesh.rst:92]:
```suggestion

For devices with a single Network Port the commands are:

```

File: docs/features/wired-mesh.rst (hunk 7 above)
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:114]: I'm not sure about it either.
I felt like it wasn't stressed enough before.

File: docs/features/wired-mesh.rst (hunk 8 above)
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:92]: I added them lower case.

Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:92]: 4a2c4dbc675f37c9fcabe5e30fd9e75286a9ee98

File: docs/features/wired-mesh.rst [hunk 9]
```diff
@@ -56,32 +56,63 @@ the ``mesh`` role to the ``interfaces.*.default_roles`` options in site.conf.
...
+Be careful: When the client role is assigned to an interface at the same time as other roles (like 'client', 'mesh' in the Mesh-on-LAN example below), the other roles take precedence (enabling mesh, but not client in said example).
+
+The setup/config-mode interface is every interface with the role ``client``, which makes removing it from interfaces not only unnecessary, but generally unrecommended.
+
+In order to make persistent changes to the routers configuration it's necessary to:
+
+* change the sections in ``/etc/config/gluon`` e.g. using uci (see examples below)
+* call ``gluon-reconfigure`` to regenerate ``/etc/config/network`` afterwards
+* apply the networking changes, either through executing ``service network restart`` or by performing a ``reboot`` of the device
```
[the remark this answers is not in the document]
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst]: removed in 4a2c4dbc675f37c9fcabe5e30fd9e75286a9ee98

File: docs/features/wired-mesh.rst (hunk 8 above)
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:92]: I'm not sure I like that too much.
But the idea very. Maybe I'd reword it to get rid of the colon, as it doubles in the next line.

`For devices with a single network port, instead of `iface_lan` and `iface_wan` configuration is done on `iface_single` as it's more specific.

Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:92]: > For single-port-devices, instead of `iface_lan` and `iface_wan` configuration is done in `iface_single`.

Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:92]: f988863c4403d6f8d809cbf9cfdd86f70c0db904

**Review by herbetom — APPROVED** [review]

**Comment by AiyionPrime** [comment]
:heavy_check_mark: img **Lint / Lua (pull_request)** *Successful on aiyions laptop*

Opposing the `Lint / Lua`-action I just confirmed that `make lint`returns without errors.
The action just has trouble talking to azure again.

**Comment by rotanid** [comment]
this looks like the wiki page also needs an update for Gluon >=2022.x ?
https://github.com/freifunk-gluon/gluon/wiki/Commandline-administration

**Comment by herbetom** [comment]
@rotanid I've updated the Wiki a bit, but my motivation isn't that huge to fix everything and it should probably be discussed wheter it's worth keeping stuff in that reference that's kompletly outdated. 

https://github.com/freifunk-gluon/gluon/wiki/Commandline-administration/_compare/7c5251ea6ef01c63aca7469ef4d8e3d65bd654f5...ade9b8716fe18e83e0a9fbd38ce5423bd0b5730f

**Comment by rotanid** [comment]
@herbetom we can link to the Gluon docs to avoid duplicate content - e.g. iface single would be missing atm if we don't link, so maybe a link would be best in sections that are covered by our docs?

**Comment by AiyionPrime** [comment]
suqashed the commit of @herbetom and @blocktrron

File: docs/features/wired-mesh.rst [hunk 10]
```diff
@@ -51,37 +51,81 @@ Both Mesh-on-WAN and Mesh-on-LAN can be configured on the "Network" page
...
+Every interface has a list of roles assigned to it which can be ``client``, ``mesh`` or ``uplink``.
+
+When the client role is assigned to an interface in combination with other roles 
+(like 'client', 'mesh' in the Mesh-on-LAN example below), the other roles take
+precedence, enabling mesh but not client in the previous example.
+
+The setup/config-mode interface is every interface with the role ``client`` which makes removing
+it from interfaces not only unnecessary, but generally unrecommended.
+
+In order to make persistent changes to the routers configuration it's necessary to:
```
Comment (rotanid) [inline docs/features/wired-mesh.rst:74]: router's

File: docs/features/wired-mesh.rst [hunk 11]
```diff
@@ -51,37 +51,81 @@ Both Mesh-on-WAN and Mesh-on-LAN can be configured on the "Network" page
...
+  uci add_list gluon.iface_single.role='mesh'
+  uci commit gluon
+
+Disable Mesh-on-Single::
+
+  uci del_list gluon.iface_single.role='mesh'
+  uci commit gluon
+
+Furthermore it is possible to make use of 802.1Q VLAN.
+The following statements would create a VLAN with id 8 on ``eth0`` and join the mesh network with it::
```
Comment (rotanid) [inline docs/features/wired-mesh.rst:114]: double :: at the end of line?

File: docs/features/wired-mesh.rst [hunk 12]
```diff
@@ -51,37 +51,81 @@ Both Mesh-on-WAN and Mesh-on-LAN can be configured on the "Network" page
...
+
+  uci set gluon.iface_lan_vlan8=interface
+  uci set gluon.iface_lan_vlan8.name='eth0.8'
+  uci add_list gluon.iface_lan_vlan8.role='mesh'
+  uci commit gluon
+
+Other VLAN-interfaces could be configured on the same parent interface in order to have
+all three roles avaliable on ``eth0`` without having them interfere with each other.
```
Comment (rotanid) [inline docs/features/wired-mesh.rst:122]:
```suggestion
all three roles available on ``eth0`` without having them interfere with each other.
```

File: docs/features/wired-mesh.rst (hunk 11 above)
Comment (rotanid) [inline docs/features/wired-mesh.rst:114]:
```suggestion
The following statements would create a VLAN with id 8 on ``eth0`` and join the mesh network with it:
```

File: docs/features/wired-mesh.rst (hunk 10 above)
Comment (rotanid) [inline docs/features/wired-mesh.rst:74]:
```suggestion
In order to make persistent changes to the router's configuration it's necessary to:
```

File: docs/features/wired-mesh.rst [hunk 13]
```diff
@@ -51,37 +51,81 @@ Both Mesh-on-WAN and Mesh-on-LAN can be configured on the "Network" page
...
+This feature comes in especially handy for the persistent configuration of virtual machines
```
Comment (rotanid) [inline docs/features/wired-mesh.rst:123]:
```suggestion
This feature comes in handy especially for the persistent configuration of virtual machines
```
but i'm not too sure about this change

File: docs/features/wired-mesh.rst [hunk 14]
```diff
@@ -51,37 +51,81 @@ Both Mesh-on-WAN and Mesh-on-LAN can be configured on the "Network" page
...
+as offloader for bigger installations.
+
+A ``reboot`` is not sufficient to apply altered configuration, a previous ``gluon-reconfigure``
```
Comment (rotanid) [inline docs/features/wired-mesh.rst:126]:
```suggestion
A ``reboot`` is not sufficient to apply an altered configuration, a prior ``gluon-reconfigure``
```

File: docs/features/wired-mesh.rst (hunk 12 above)
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:122]: nice catch, thanks.

File: docs/features/wired-mesh.rst (hunk 11 above)
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:114]: The double-double colon is necessary for the following code block to be formatted properly. Can't commit this.

File: docs/features/wired-mesh.rst (hunk 10 above)
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:74]: nice, thanks

File: docs/features/wired-mesh.rst (hunk 13 above)
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:123]: They are both valid english;
I meant what I wrote, the changed variant changes the meaning of the sentence and feels like a weird shoulder pat to ourselves, to me.

File: docs/features/wired-mesh.rst (hunk 14 above)
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst:126]: reworded it a little:

A ``reboot`` is not sufficient to apply an altered configuration, calling ``gluon-reconfigure``
before is mandatory in order for changes to take effect.

File: docs/features/wired-mesh.rst (hunk 12 above)
Comment (AiyionPrime) [reply docs/features/wired-mesh.rst]: 714e2e3439eaae202a4cf78a520a30f8c02f2605

**Comment by AiyionPrime** [comment]
Squashed @rotanid's fixes as well.
As @herbetom and @blocktrron reviewed the contents earlier, this should be ready to merge.
